File: generate_item/generate_item/report/valve_testing_report/valve_testing_report.py
# Copyright (c) 2026, Finbyz and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
	"""Return columns, data, message, chart, and report_summary for Valve Testing Report."""
	filters = frappe._dict(filters or {})
	columns = get_columns()
	data = get_data(filters)
	report_summary = get_report_summary(data)
	chart = get_chart(data)

	message = None
	if not data:
		message = _("No records found for the selected filters.")

	return columns, data, message, chart, report_summary


def get_columns() -> list[dict]:
	"""Return column definitions strictly based on Valve Testing & Valve Testing Item schema."""
	return [
		{
			"label": _("Valve Testing"),
			"fieldname": "valve_testing",
			"fieldtype": "Link",
			"options": "Valve Testing",
			"width": 140,
		},
		{
			"label": _("Posting Date"),
			"fieldname": "posting_date",
			"fieldtype": "Date",
			"width": 100,
		},
		{
			"label": _("Test Date"),
			"fieldname": "test_date",
			"fieldtype": "Date",
			"width": 100,
		},
		{
			"label": _("Branch"),
			"fieldname": "branch",
			"fieldtype": "Link",
			"options": "Branch",
			"width": 100,
		},
		{
			"label": _("Inspector"),
			"fieldname": "user",
			"fieldtype": "Link",
			"options": "Employee",
			"width": 110,
		},
		{
			"label": _("Inspector Name"),
			"fieldname": "user_name",
			"fieldtype": "Data",
			"width": 140,
		},
		{
			"label": _("Valve Sr No"),
			"fieldname": "serial_number",
			"fieldtype": "Link",
			"options": "Serial Number",
			"width": 130,
		},
		{
			"label": _("Item Code"),
			"fieldname": "item_code",
			"fieldtype": "Link",
			"options": "Item",
			"width": 140,
		},
		{
			"label": _("Item Name"),
			"fieldname": "item_name",
			"fieldtype": "Data",
			"width": 160,
		},
		{
			"label": _("Sales Order"),
			"fieldname": "sales_order",
			"fieldtype": "Link",
			"options": "Sales Order",
			"width": 130,
		},
		{
			"label": _("Batch No"),
			"fieldname": "batch_no",
			"fieldtype": "Link",
			"options": "Batch",
			"width": 120,
		},
		{
			"label": _("Valve Type"),
			"fieldname": "type",
			"fieldtype": "Data",
			"width": 110,
		},
		{
			"label": _("Size"),
			"fieldname": "size",
			"fieldtype": "Data",
			"width": 80,
		},
		{
			"label": _("Class/Rating"),
			"fieldname": "class",
			"fieldtype": "Data",
			"width": 90,
		},
		{
			"label": _("MOC"),
			"fieldname": "shell_moc",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": _("End Connection"),
			"fieldname": "end_connection",
			"fieldtype": "Data",
			"width": 110,
		},
		{
			"label": _("Operation"),
			"fieldname": "operation",
			"fieldtype": "Data",
			"width": 100,
		},
		{
			"label": _("Inch Factor"),
			"fieldname": "inch_factor",
			"fieldtype": "Data",
			"width": 90,
		},
		{
			"label": _("Size Value"),
			"fieldname": "value",
			"fieldtype": "Data",
			"width": 80,
		},
		{
			"label": _("Inspector Inches"),
			"fieldname": "inches",
			"fieldtype": "Float",
			"width": 110,
		},
		{
			"label": _("Test Result"),
			"fieldname": "test_ok",
			"fieldtype": "Data",
			"width": 110,
		},
		{
			"label": _("Leak Reason"),
			"fieldname": "leak_reason",
			"fieldtype": "Data",
			"width": 140,
		},
		{
			"label": _("Remark"),
			"fieldname": "remark",
			"fieldtype": "Data",
			"width": 160,
		},
	]


def get_data(filters: dict) -> list[dict]:
	"""Fetch data by joining Valve Testing (parent) with Valve Testing Item (child), submitted only."""
	conditions, values = build_conditions(filters)

	sql = f"""
		SELECT
			vt.name AS valve_testing,
			vt.posting_date AS posting_date,
			vti.date AS test_date,
			vt.branch AS branch,
			vt.user AS user,
			emp.employee_name AS user_name,
			vt.docstatus AS docstatus,
			vti.serial_number AS serial_number,
			vti.item_code AS item_code,
			item.item_name AS item_name,
			vti.sales_order AS sales_order,
			vti.batch_no AS batch_no,
			vti.type AS type,
			vti.size AS size,
			vti.class AS class,
			vti.shell_moc AS shell_moc,
			vti.end_connection AS end_connection,
			vti.operation AS operation,
			vti.inch_factor AS inch_factor,
			vti.value AS value,
			vti.inches AS inches,
			vti.test_ok AS test_ok,
			vti.leak_reason AS leak_reason,
			vti.remark AS remark
		FROM `tabValve Testing` vt
		INNER JOIN `tabValve Testing Item` vti
			ON vti.parent = vt.name AND vti.parenttype = 'Valve Testing'
		LEFT JOIN `tabEmployee` emp
			ON emp.name = vt.user
		LEFT JOIN `tabItem` item
			ON item.name = vti.item_code
		WHERE vt.docstatus = 1
			{conditions}
		ORDER BY vt.posting_date DESC, vt.name DESC, vti.idx ASC
	"""

	data = frappe.db.sql(sql, values, as_dict=True)

	for row in data:
		if row.inches:
			row.inches = flt(row.inches)

	return data


def build_conditions(filters: dict) -> tuple[str, dict]:
	"""Construct SQL WHERE conditions securely using parameterized queries."""
	conditions = []
	values = {}

	if filters.get("from_date"):
		conditions.append("vt.posting_date >= %(from_date)s")
		values["from_date"] = filters.get("from_date")

	if filters.get("to_date"):
		conditions.append("vt.posting_date <= %(to_date)s")
		values["to_date"] = filters.get("to_date")

	if filters.get("branch"):
		conditions.append("vt.branch = %(branch)s")
		values["branch"] = filters.get("branch")

	if filters.get("valve_testing"):
		conditions.append("vt.name = %(valve_testing)s")
		values["valve_testing"] = filters.get("valve_testing")

	if filters.get("user"):
		conditions.append("vt.user = %(user)s")
		values["user"] = filters.get("user")

	if filters.get("sales_order"):
		conditions.append("vti.sales_order = %(sales_order)s")
		values["sales_order"] = filters.get("sales_order")

	if filters.get("batch_no"):
		conditions.append("vti.batch_no = %(batch_no)s")
		values["batch_no"] = filters.get("batch_no")

	if filters.get("serial_number"):
		conditions.append("vti.serial_number = %(serial_number)s")
		values["serial_number"] = filters.get("serial_number")

	if filters.get("item_code"):
		conditions.append("vti.item_code = %(item_code)s")
		values["item_code"] = filters.get("item_code")

	if filters.get("test_ok") and filters.get("test_ok") not in ("Both", "All", ""):
		conditions.append("vti.test_ok = %(test_ok)s")
		values["test_ok"] = filters.get("test_ok")

	if filters.get("leak_reason"):
		conditions.append("vti.leak_reason = %(leak_reason)s")
		values["leak_reason"] = filters.get("leak_reason")

	cond_str = ("AND " + " AND ".join(conditions)) if conditions else ""
	return cond_str, values


def get_report_summary(data: list[dict]) -> list[dict]:
	"""Generate executive KPI summary cards for Valve Testing Report."""
	if not data:
		return []

	total_tested = len(data)
	no_leak_accepted = sum(1 for r in data if r.get("test_ok") == "Accepted")
	not_accepted = sum(1 for r in data if r.get("test_ok") == "Not Accepted")
	total_inches = sum(flt(r.get("inches")) for r in data if r.get("inches"))

	acceptance_rate = (no_leak_accepted / total_tested * 100) if total_tested else 0.0

	return [
		{
			"value": total_tested,
			"label": _("Total Tested Valves"),
			"datatype": "Int",
			"indicator": "blue",
		},
		{
			"value": no_leak_accepted,
			"label": _("Accepted"),
			"datatype": "Int",
			"indicator": "green",
		},
		{
			"value": not_accepted,
			"label": _("Not Accepted"),
			"datatype": "Int",
			"indicator": "red" if not_accepted > 0 else "gray",
		},
		{
			"value": f"{acceptance_rate:.1f}%",
			"label": _("Acceptance Rate"),
			"datatype": "Data",
			"indicator": "green" if acceptance_rate >= 90 else ("orange" if acceptance_rate >= 75 else "red"),
		},
		{
			"value": total_inches,
			"label": _("Total Inspector Inches"),
			"datatype": "Float",
			"indicator": "purple",
		},
	]


def get_chart(data: list[dict]) -> dict | None:
	"""Generate a clean Bar Chart showing Not Accepted (Red) and Accepted (Green) bars."""
	if not data:
		return None

	not_accepted = sum(1 for r in data if r.get("test_ok") == "Not Accepted")
	accepted = sum(1 for r in data if r.get("test_ok") == "Accepted")

	if not_accepted == 0 and accepted == 0:
		return None

	return {
		"data": {
			"labels": [_("Valves Status")],
			"datasets": [
				{
					"name": _("Not Accepted"),
					"values": [not_accepted],
				},
				{
					"name": _("Accepted"),
					"values": [accepted],
				},
			],
		},
		"type": "bar",
		"height": 220,
		"colors": ["#ff5858", "#28a745"],
	}

File: generate_item/generate_item/report/valve_testing_report/valve_testing_report.js
// Copyright (c) 2026, Finbyz and contributors
// For license information, please see license.txt

function format_iso_date(d) {
	let year = d.getFullYear();
	let month = String(d.getMonth() + 1).padStart(2, "0");
	let day = String(d.getDate()).padStart(2, "0");
	return `${year}-${month}-${day}`;
}

function calculate_date_range(range_type) {
	let now = new Date();
	let year = now.getFullYear();
	let month = now.getMonth();
	let day = now.getDate();

	if (range_type === "Today") {
		let d = format_iso_date(now);
		return [d, d];
	} else if (range_type === "Yesterday") {
		let y = new Date(year, month, day - 1);
		let d = format_iso_date(y);
		return [d, d];
	} else if (range_type === "This Week") {
		let dayOfWeek = now.getDay(); // 0 is Sunday, 1 is Monday...
		let diffToMon = (dayOfWeek === 0 ? -6 : 1) - dayOfWeek;
		let monday = new Date(year, month, day + diffToMon);
		let sunday = new Date(monday.getFullYear(), monday.getMonth(), monday.getDate() + 6);
		return [format_iso_date(monday), format_iso_date(sunday)];
	} else if (range_type === "Last Week") {
		let dayOfWeek = now.getDay();
		let diffToMon = (dayOfWeek === 0 ? -6 : 1) - dayOfWeek;
		let lastMonday = new Date(year, month, day + diffToMon - 7);
		let lastSunday = new Date(lastMonday.getFullYear(), lastMonday.getMonth(), lastMonday.getDate() + 6);
		return [format_iso_date(lastMonday), format_iso_date(lastSunday)];
	} else if (range_type === "This Month") {
		let firstDay = new Date(year, month, 1);
		let lastDay = new Date(year, month + 1, 0);
		return [format_iso_date(firstDay), format_iso_date(lastDay)];
	} else if (range_type === "Last Month") {
		let firstDay = new Date(year, month - 1, 1);
		let lastDay = new Date(year, month, 0);
		return [format_iso_date(firstDay), format_iso_date(lastDay)];
	}
	return null;
}

frappe.query_reports["Valve Testing Report"] = {
	filters: [
		{
			fieldname: "date_range",
			label: __("Date Range"),
			fieldtype: "Select",
			options: [
				"Today",
				"Yesterday",
				"This Week",
				"Last Week",
				"This Month",
				"Last Month",
				"Custom",
			],
			default: "This Month",
			on_change: function (query_report) {
				let range = query_report.get_filter_value("date_range");
				if (range && range !== "Custom") {
					let dates = calculate_date_range(range);
					if (dates) {
						query_report.set_filter_value("from_date", dates[0]);
						query_report.set_filter_value("to_date", dates[1]);
					}
				}
			},
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: calculate_date_range("This Month")[0],
			reqd: 1,
			on_change: function (query_report) {
				let range = query_report.get_filter_value("date_range");
				if (range && range !== "Custom") {
					let expected = calculate_date_range(range);
					if (expected && expected[0] !== query_report.get_filter_value("from_date")) {
						query_report.set_filter_value("date_range", "Custom");
					}
				}
			},
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: calculate_date_range("This Month")[1],
			reqd: 1,
			on_change: function (query_report) {
				let range = query_report.get_filter_value("date_range");
				if (range && range !== "Custom") {
					let expected = calculate_date_range(range);
					if (expected && expected[1] !== query_report.get_filter_value("to_date")) {
						query_report.set_filter_value("date_range", "Custom");
					}
				}
			},
		},
		{
			fieldname: "test_ok",
			label: __("Test Result"),
			fieldtype: "Select",
			options: "Accepted\nNot Accepted\nBoth",
			default: "Accepted",
		},
		{
			fieldname: "testing_phase",
			label: __("Testing Phase"),
			fieldtype: "Select",
			options: "\nPre Testing\nFinal Testing\nTPI Testing",
		},
		{
			fieldname: "branch",
			label: __("Branch"),
			fieldtype: "Link",
			options: "Branch",
		},
		{
			fieldname: "leak_reason",
			label: __("Leak Reason"),
			fieldtype: "Select",
			options: "\nAir Seat Leak\nHydro Seat Leak\nBody Casting Leak\nSeatring Casting Leak\nBonnet Casting Leak\nWedge Casting Leak\nJoint Leak\nBack Seat Leak\nWelding Leak\nBore Casting Leak",
		},
		{
			fieldname: "valve_testing",
			label: __("Valve Testing"),
			fieldtype: "Link",
			options: "Valve Testing",
		},
		{
			fieldname: "sales_order",
			label: __("Sales Order"),
			fieldtype: "Link",
			options: "Sales Order",
		},
		{
			fieldname: "batch_no",
			label: __("Batch No"),
			fieldtype: "Link",
			options: "Batch",
		},
		{
			fieldname: "serial_number",
			label: __("Valve Sr No"),
			fieldtype: "Link",
			options: "Serial Number",
		},
		{
			fieldname: "item_code",
			label: __("Item Code"),
			fieldtype: "Link",
			options: "Item",
		},
		{
			fieldname: "user",
			label: __("Inspector"),
			fieldtype: "Link",
			options: "Employee",
		},
		{
			fieldname: "tested_by",
			label: __("Tested By"),
			fieldtype: "Link",
			options: "User",
			get_query: function () {
				return {
					filters: {
						enabled: 1,
						custom_is_operator: 1,
					},
				};
			},
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "test_ok" && data) {
			if (data.test_ok === "Accepted") {
				value = `<span style="background: rgba(16, 185, 129, 0.14); color: #059669; border: 1px solid rgba(16, 185, 129, 0.35); font-weight: 600; padding: 3px 10px; border-radius: 12px; font-size: 11px; display: inline-flex; align-items: center; gap: 4px;">✓ ${__("Accepted")}</span>`;
			} else if (data.test_ok === "Not Accepted") {
				value = `<span style="background: rgba(239, 68, 68, 0.14); color: #dc2626; border: 1px solid rgba(239, 68, 68, 0.35); font-weight: 600; padding: 3px 10px; border-radius: 12px; font-size: 11px; display: inline-flex; align-items: center; gap: 4px;">✗ ${__("Not Accepted")}</span>`;
			}
		}

		if (column.fieldname === "leak_reason" && data) {
			if (data.test_ok === "Not Accepted" && data.leak_reason) {
				value = `<span style="background: rgba(244, 63, 94, 0.12); color: #e11d48; border: 1px solid rgba(244, 63, 94, 0.3); padding: 2px 8px; border-radius: 6px; font-weight: 600; font-size: 11px; display: inline-flex; align-items: center; gap: 4px;">⚠️ ${data.leak_reason}</span>`;
			} else if (data.test_ok === "Accepted") {
				value = "";
			}
		}

		if (column.fieldname === "serial_number" && data && data.serial_number) {
			value = `<span style="font-family: var(--font-stack-monospace, monospace); font-weight: 600; color: #2563eb;">${value}</span>`;
		}

		if (column.fieldname === "inches" && data && data.inches) {
			value = `<span style="font-weight: 600; color: #7c3aed;">${value}</span>`;
		}

		return value;
	},
};
